### PR TITLE
UI changes for column ids in records API

### PR DESCRIPTION
**Pull request by seancolsen**
These are the front end changes for #1047.

I put them in a separate PR to aid review. I figure we can merge this first (into `#1047`), then merge `#1047`.

## Issues

- Grouping is broken because the response from the records API is still using column names within the `grouping` properties.

    Here's the GET response with grouping.

    ```json5
    {
      count: 10,
      grouping: {
        columns: ["country"], // <-- column name here
        mode: "distinct",
        num_groups: null,
        ranged: false,
        groups: [
          {
            count: 1,
            first_value: { country: "Bhutan" }, // <-- column name here
            last_value: { country: "Bhutan" }, // <-- column name here
            result_indices: [0],
          },
          // ...
        ],
      },
      results: [
        //...
      ],
    }
    ```

    I'm not sure how we want to handle this. I _thought_ I tested this when I reviewed `#1047` initially, but it may have slipped through the cracks. Or it may be a regression from changes pushed to that PR more recently. @silentninja any thoughts?

    I'd lean towards getting all of this work merged into `master` (thereby breaking the grouping feature temporarily) and fixing it with subsequent issues/PRs. My understanding is that some of the grouping API stuff is about to change anyway.

    We don't have any E2E tests for grouping yet which is why you don't see test failures.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: mathesar_ui/src/stores/table-data/display.ts
Comment (seancolsen) [inline mathesar_ui/src/stores/table-data/display.ts:43]: @pavish Would you like to lend a hand at clarifying this design? Maybe you could push a commit that cleans this up a bit by setting a better name for this constant and re-writing the comments above it? I also considered allowing strings as keys in the `ColumnPositionMap` so that we could keep the previous value of `__row`, but `-1` seemed a bit cleaner to me.

Comment (pavish) [reply mathesar_ui/src/stores/table-data/display.ts:43]: Each entry within the columnPositionMap corresponds to the position of the column. However, the `_row` value consists of the total width & left values (i.e the position) of the row. It's placed within columnPositionMap inorder to avoid calculation within the component, which will run for each row.

We could use a dedicated store for it or even a new class containing both columnPosition and row width.

It was placed within `ColumnPositionMap` because we were passing around a lot of props to the child components and I wanted to reduce the number of props. Now, it is all passed down using context, so that's no longer an issue.

**Comment by mathemancer** [comment]
> * I'd lean towards getting all of this work merged into `master` (thereby breaking the grouping feature temporarily) and fixing it with subsequent issues/PRs. My understanding is that some of the grouping API stuff is about to change anyway.

I would support that.  I'm currently blocked on grouping until the column ID stuff is merged, and it might be more efficient for me to change it as part of the next grouping PR instead.

**Comment by kgodey** [comment]
I agree re: getting it merged sooner and breaking grouping temporarily.

File: mathesar_ui/src/stores/table-data/display.ts
Comment (seancolsen) [reply mathesar_ui/src/stores/table-data/display.ts:43]: Thanks for the explanation, @pavish. Are my changes in this PR with respect to `ColumnPositionMap` okay for now? Or would you like further adjustment? If you'd like to see more clean up here, can you either add a commit or let me know how you'd like me to address it?

Comment (pavish) [reply mathesar_ui/src/stores/table-data/display.ts:43]: @seancolsen I think the change is fine. I wanted to specify it since you weren't sure about the context. You could just update your comment in this PR.

**Comment by silentninja** [comment]
@seancolsen I agree on merging along with breaking changes. I created a discussion to decide on [Which column identifier should we use for DML operation](https://github.com/centerofci/mathesar/discussions/1147). There are a few conflicting thoughts and there could be a possibility of using `column name`

**Review by pavish — APPROVED** [review]
Looks good to me, I didn't test it thoroughly and I didn't see any breaking functionality so far (apart from grouping, which is mentioned in the PR description).

**Pull request merged**
